resource-aggregate/cqrs/eventstore/mongodb/load.go

```diff
@@ -309,26 +322,134 @@ func snapshotQueriesToMongoQuery(groupID string, queries []eventstore.SnapshotQu
 	if len(queries) == 0 {
 		opts.SetHint(groupIDQueryIndex)
-		return bson.D{
-			{Key: groupIDKey, Value: groupID}, {Key: isActiveKey, Value: true},
-		}, opts
+		return []mongoQuery{
+			{
+				filter: bson.D{
+					{Key: groupIDKey, Value: groupID}, {Key: isActiveKey, Value: true},
+				},
+				options: opts,
+			},
+		}
 	}
 
-	opts.SetHint(groupIDaggregateIDQueryIndex)
-	orQueries := make([]bson.D, 0, 32)
+	optsTypes := options.Find()
+	// create a copy of the options
+	*optsTypes = *opts
+	resourceQueries := make([]bson.D, 0, 32)
+	typeQueries := make([]bson.D, 0, 32)
 	for _, q := range queries {
+		if q.AggregateID == "" && len(q.Types) == 0 {
+			opts.SetHint(groupIDQueryIndex)
+			return []mongoQuery{
+				{
+					filter: bson.D{
+						{Key: groupIDKey, Value: groupID}, {Key: isActiveKey, Value: true},
+					},
+					options: opts,
+				},
+			}
+		}
+		if q.AggregateID == "" && len(q.Types) > 0 {
+			optsTypes.SetHint(groupIDTypesQueryIndex)
+			typeQueries = append(typeQueries, bson.D{{Key: groupIDKey, Value: groupID}, {Key: typesKey, Value: bson.M{"$all": q.Types}}, {Key: isActiveKey, Value: true}})
+			continue
+		}
+
 		if q.AggregateID != "" {
-			orQueries = append(orQueries, bson.D{{Key: groupIDKey, Value: groupID}, {Key: aggregateIDKey, Value: q.AggregateID}, {Key: isActiveKey, Value: true}})
+			opts.SetHint(groupIDaggregateIDQueryIndex)
+			resourceQueries = append(resourceQueries, bson.D{{Key: groupIDKey, Value: groupID}, {Key: aggregateIDKey, Value: q.AggregateID}, {Key: isActiveKey, Value: true}})
 		}
 	}
-	return bson.M{
-		"$or": orQueries,
-	}, opts
+
+	r := make([]mongoQuery, 0, 2)
+	if len(resourceQueries) > 0 {
+		r = append(r, mongoQuery{
+			filter:  bson.M{"$or": resourceQueries},
+			options: opts,
+		})
+	}
+	if len(typeQueries) > 0 {
+		r = append(r, mongoQuery{
+			filter:  bson.M{"$or": typeQueries},
+			options: optsTypes,
+		})
+	}
+	return r
 }
 
 func (s *EventStore) loadFromSnapshot(ctx context.Context, groupID string, queries []eventstore.SnapshotQuery, eventHandler eventstore.Handler) error {
-	filter, opts := snapshotQueriesToMongoQuery(groupID, queries)
-	return s.loadEventsQuery(ctx, eventHandler, nil, filter, opts)
+	mongoQueries := snapshotQueriesToMongoQuery(groupID, queries)
+	if len(mongoQueries) > 1 {
+		return errors.New("too many types of queries")
+	}
+	return s.loadEventsQuery(ctx, eventHandler, nil, mongoQueries)
+}
+
+func uniqueQueryWithEmptyAggregateID(queries []eventstore.SnapshotQuery, query eventstore.SnapshotQuery) []eventstore.SnapshotQuery {
+	if len(query.Types) == 0 {
+		// get all events without filter
+		return []eventstore.SnapshotQuery{query}
+	}
+	for idx, q := range queries {
+		if slices.Equal(q.Types, query.Types) {
+			// get all events with the certain type, replace the queries with the more or equal general one
+			queries[idx] = query
+			queries = removeDuplicates(queries, idx, func(qa eventstore.SnapshotQuery) bool { return slices.Equal(qa.Types, query.Types) })
```

Comment:
Is order revelant? ie. should slices `["a", "b"]` and `["b", "a"]` be considered duplicates? If so, `slices.Equal` won't work, unless the slices are sorted.

Reply:
fixed